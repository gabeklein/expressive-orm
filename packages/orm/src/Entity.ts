import { Field, init, num, underscore } from './fields';
import { desc, equal, Where } from './where';

type Values<T extends Type> = {
  [K in Exclude<keyof T, keyof Type> as (
    T[K] extends Type.Class ? never : K
  )]: T[K];
}

type Compat<T extends Type> = Partial<Values<T>>;

type Insert<T extends Type> = {
  [K in keyof Omit<T, keyof Type> as (
    undefined extends T[K] ? never : (T[K] extends Type.Class<any> ? never : K)
  )]: T[K];
} & {
  [K in keyof Omit<T, keyof Type> as (
    undefined extends T[K] ? (T[K] extends Type.Class<any> ? never : K) : never
  )]?: T[K];
};

declare namespace Type {
  type Class<T extends Type = Type> = (abstract new (...args: any[]) => T) & typeof Type;

  type Instance<T> = T extends Class<infer U> ? U : T;

  type KeyOf<T> = T extends Class<infer U> ? keyof U : keyof T;

  type Result<T extends Type> = PromiseLike<T[]>;

  type Conds<T> = Where | T | Conds<T>[]

  type Query<T extends {}> = {
    [K in keyof T]?: Conds<T[K]>;
  };

  type Constraint = [
    field: string,
    op: string,
    value?: any
  ]

  interface Connection {
    get: (table: string, constraints: Constraint[], limit?: number, offset?: number) => Promise<any[]>;
    insert: (table: string, data: Record<string, any>, returns?: boolean) => Promise<any>;
    update: (table: string, id: string | number, data: Record<string, any>) => Promise<void>;
    remove: (table: string, id: string | number) => Promise<void>;
  }

  export {
    Class,
    Compat,
    Connection,
    Constraint,
    Insert,
    Instance,
    KeyOf,
    Query,
    Result,
    Values
  }
}

async function reject(): Promise<never> {
  throw new Error("No database connection. Please set Entity.connection to a valid Connection instance.");
}

class NotReady {
  get = reject;
  insert = reject;
  update = reject;
  remove = reject;
}

abstract class Type {
  id = num({ optional: true });

  get values() {
    const values: Record<string, any> = {};
    const type = this.constructor as typeof Type;

    for (const [key] of type.fields)
      values[key] = (this as any)[key];

    return Object.freeze(values) as Values<this>;
  };

  protected async onDelete?(): Promise<void>;

  async update(data: Type.Compat<this>) {
    if (Object.keys(data).length === 0)
      return;

    const type = this.constructor as typeof Type
    await type.connection.update(type.ref, this.id, type.into(data, true));

    for (const [key] of type.fields)
      if (key in data)
        Object.defineProperty(this, key, {
          value: (data as any)[key],
          configurable: true,
          enumerable: true,
          writable: false,
        });
  }

  async delete() {
    const { connection, ref } = this.constructor as typeof Type;

    if (typeof this.onDelete === "function")
      await this.onDelete();

    await connection.remove(ref, this.id);
  }

  static schema?: string;
  static subset?: Record<string, unknown>;
  static connection: Type.Connection = new NotReady();
  static table?: string;

  static get ref() {
    const { name, schema, table = underscore(name) } = this;
    const ref = schema ? `${schema}.${table}` : table;
    Object.defineProperty(this, 'ref', { value: ref });
    return ref;
  }

  static get fields(): Map<string, Field> {
    const fields = init(this);
    Object.defineProperty(this, 'fields', { value: fields });
    return fields;
  }

  protected static from<T extends Type>(
    this: Type.Class<T>,
    raw: Record<string, any>
  ): T {
    const entity = Object.create(this.prototype) as T;

    this.fields.forEach(field => {
      Object.assign(entity, field.input(raw));
    });

    return entity;
  }

  protected static into<T extends Type>(
    this: Type.Class<T>,
    data: Compat<T>,
    partial?: boolean
  ): Record<string, any> {
    const result: Record<string, any> = {};

    this.fields.forEach(field => {
      Object.assign(result, field.output(data, partial));
    })

    return result;
  }

  static async set<T extends Type>(this: Type.Class<T>, id: string, data: Type.Compat<T>) {
    await this.connection.update(this.ref, id, this.into(data, true));
  }

  static async new<T extends Type>(this: Type.Class<T>, data: Type.Insert<T>): Promise<T>;
  static async new<T extends Type>(this: Type.Class<T>, data: Type.Insert<T>[]): Promise<void>;
  static async new<T extends Type, U>(this: Type.Class<T>, data: Array<U>, mapFn: (item: U) => Type.Insert<T>): Promise<void>;
  static async new<T extends Type>(
    this: Type.Class<T>,
    data: Values<T> | unknown[],
    mapFn?: (item: unknown) => any
  ) {
    const insert = async (d: Values<T>, returns?: boolean) => {
      const input = this.into({ ...d as any, ...this.subset });
      const inserted = await this.connection.insert(this.ref, input, returns);
      return this.from(inserted) as T;
    };

    if (Array.isArray(data)) {
      await Promise.all(data.map(mapFn || (x => x)).map(x => insert(x)))
      return;
    }

    return insert(data, true);
  }

  static async one<T extends Type>(this: Type.Class<T>, id: number, expect?: true): Promise<T>;
  static async one<T extends Type>(this: Type.Class<T>, id: number, expect: boolean): Promise<T | undefined>;
  static async one<T extends Type>(this: Type.Class<T>, where: Type.Query<T>, expect: boolean): Promise<T | undefined>;
  static async one<T extends Type>(this: Type.Class<T>, where?: Type.Query<T>, expect?: true): Promise<T>;
  static async one<T extends Type>(this: Type.Class<T>, where?: number | Type.Query<T>, expect?: boolean): Promise<T | undefined> {
    if (typeof where === "number")
      where = { id: equal(where) };

    const [row] = await this.query(where, 1);

    if (!row && expect !== false)
      throw new Error(`No ${this.name} found with: \n${JSON.stringify(where, null, 2)}\n`);

    return row;
  }

  static async query<T extends Type>(
    from?: Type.Constraint[] | Type.Query<T>,
    limit?: number,
    offset: number = 0
  ): Promise<T[]> {
    const { connection, subset, ref } = this;

    if (!from)
      from = { id: desc() };

    if (!Array.isArray(from))
      from = buildConstraints({ ...from, ...subset });

    from = from.map(([key, op, value]) => {
      const field = this.fields.get(key);

      if(field)
        return [field.column, op, field.set(value)]
    }).filter(Boolean) as Type.Constraint[];

    const rows = await connection.get(ref, from, limit, offset);
    return rows.map(row => this.from(row) as T)
  }

  static async get<T extends Type>(
    this: Type.Class<T>,
    where?: Type.Query<T>,
    limit?: number,
    offset?: number
  ) {
    return this.query(where, limit, offset);
  }

  static async *all<T extends Type>(
    this: Type.Class<T>,
    pageSize: number,
    where?: Type.Query<T>
  ): AsyncGenerator<T[]> {
    let offset = 0;
    let done = false;

    while (!done) {
      const batch = await this.query(where, pageSize, offset);

      offset += batch.length;

      if (!batch.length) break;
      if (pageSize && batch.length < pageSize) done = true;

      yield batch;
    }
  }
}

function buildConstraints(where: Type.Query<any>) {
  const constraints: Type.Constraint[] = [];

  for (const key in where) {
    let field = where[key] as Where | Where[];

    if (!Array.isArray(field))
      field = [field];

    for (const val of field) {
      if (val === undefined || val instanceof Where && val.skip)
        continue;

      constraints.push(
        val instanceof Where
          ? [key, val.op || "=", val.value]
          : [key, '=', val]
      )
    }
  }

  return constraints;
}

export {
  Type,
  Type as default,
  Values,
  Compat,
  Insert,
};