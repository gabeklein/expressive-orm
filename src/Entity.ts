import Connection from './connection/Connection';
import Field from './Field';
import Primary from './fields/Primary';
import Query from './query/Query';

export type InstanceOf<T> = T extends { prototype: infer U } ? U : never;

const REGISTER = new Set<Entity.Type>();
const INSTRUCTION = new Map<symbol, Entity.Instruction>();

const describe = Object.getOwnPropertyDescriptor;
const define = Object.defineProperty;

declare namespace Entity {
  type Type<T extends Entity = Entity> =
    & (abstract new () => T)
    & typeof Entity;

  type List<T extends Entity> = T[];

  type DataRecusive<T> =
    T extends string ? string :
    T extends number ? number :
    T extends boolean ? boolean :
    T extends List<infer U> ? Pure<U>[] :
    T extends Entity ? Pure<T> : T;

  type Pure<T extends Entity> = {
    [K in Field<T>]-?: DataRecusive<T[K]>;
  }

  type Field<T extends Entity> = Exclude<keyof T, "table">;

  type Where<T extends Entity, R> =
    (source: Query.Type<T>, query: Query.Where) => () => R;

  type Instruction = (parent: Entity.Type, key: string) => void;
}

abstract class Entity {
  this!: Entity.Type;
  id = Primary();

  static table: string;
  static schema: string;
  static fields: Map<string, Field>;
  static deps: Set<Entity.Type>;
  static connection?: Connection;
  static focus?: { [key: string]: any };

  static ensure<T extends Entity>(
    this: Entity.Type<T>,
    connection?: Connection){

    if(!REGISTER.has(this)){
      REGISTER.add(this);

      this.table = /class (\w+?) /.exec(this.toString())![1];
      this.schema = "";
      this.connection = connection;
      this.fields = new Map();
      this.deps = new Set();
      
      const prime = new (this as any)();
      
      for(const key in prime){
        const { value } = describe(prime, key)!;
        const instruction = INSTRUCTION.get(value);    
  
        if(!instruction)
          continue;
  
        INSTRUCTION.delete(value);
        instruction(this, key);
        define(prime, key, {
          get: () => this.focus![key],
          set: is => this.focus![key] = is
        })
      }
    }

    return this;
  }

  static field(instruction: Entity.Instruction){
    const placeholder = Symbol(`ORM instruction`);
    INSTRUCTION.set(placeholder, instruction);
    return placeholder as any;
  }

  /**
   * Create an arbitary map of managed fields.
   */
  static map<T extends Entity>(
    this: Entity.Type<T>,
    getValue: (type: Field, key: Entity.Field<T>, proxy: {}) => any,
    cache?: boolean
  ){
    const proxy = {} as any;

    this.ensure();

    for(const [key, type] of this.fields)
      define(proxy, key, {
        configurable: true,
        get: () => {
          const value = getValue.call(
            proxy,
            type,
            key as Entity.Field<T>,
            proxy
          );

          if(cache !== false)
            define(proxy, key, { value });

          return value;
        }
      })
    
    return proxy;
  }

  // static select<T extends Entity, R>(
  //   this: Entity.Type<T>,
  //   from: Entity.Where<T, R>
  // ){
  //   return new Query(where => {
  //     return from(where(this), where);
  //   })
  // }

  // static insert<T extends Entity, R>(
  //   this: Entity.Type<T>,
  //   data: Insert.Values<T>
  // ){
  //   return new Insert(this, data).exec();
  // }

  // static async get<T extends Entity, R>(
  //   this: Entity.Type<T>,
  //   from: Entity.Where<T, R>
  // ){
  //   return this.select(from).get();
  // }

  // static async getOne<T extends Entity, R>(
  //   this: Entity.Type<T>,
  //   from: Entity.Where<T, R>
  // ){
  //   return this.select(from).getOne();
  // }
}

export default Entity;