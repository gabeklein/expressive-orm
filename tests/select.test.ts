import Entity, { Int, One, Query, VarChar } from '../src';

class Foo extends Entity {
  bar = VarChar();
  baz = VarChar();
}

describe("where.get", () => {
  it("will select via object", () => {
    const query = new Query(where => {
      const { bar, baz } = where(Foo);
  
      return where.get({ bar, baz })
    })
  
    expect(query).toMatchInlineSnapshot(`
SELECT
  \`Foo\`.\`bar\` AS \`bar\`,
  \`Foo\`.\`baz\` AS \`baz\`
FROM
  \`Foo\`
`);
  })
  
  it("will select via map function", () => {
    const query = new Query(where => {
      const foo = where(Foo);
  
      return where.get(() => {
        return {
          bar: foo.bar,
          baz: foo.baz
        };
      });
    })
  
    expect(query).toMatchInlineSnapshot(`
SELECT
  \`Foo\`.\`bar\` AS \`1\`,
  \`Foo\`.\`baz\` AS \`2\`
FROM
  \`Foo\`
`);
  })
  
  it("will select a field directly", () => {
    const query = new Query(where => {
      const { bar } = where(Foo);
  
      return where.get(bar);
    })
  
    expect(query).toMatchInlineSnapshot(`
SELECT
  \`Foo\`.\`bar\` AS \`1\`
FROM
  \`Foo\`
`);
  })
})

describe("joins", () => {
  class Foo extends Entity {
    name = VarChar();
    color = VarChar();
  }

  it("will select joined values", async () => {
    class Bar extends Entity {
      name = VarChar();
      color = VarChar();
      rating = Int();
    }
    
    class Baz extends Entity {
      rating = Int();
    }
  
    const query = new Query(where => {
      const foo = where(Foo);
      const bar = where(Bar, { color: foo.color });
      const baz = where(Baz, "left", { rating: bar.rating });
  
      where(foo.name).not("Danny");
      where(bar.rating).greater(50);
  
      return where.get({
        fooValue: foo.name,
        barValue: bar.name,
        bazRating: baz.rating
      })
    });
  
    expect(query).toMatchInlineSnapshot(`
  SELECT
    \`Foo\`.\`name\` AS \`fooValue\`,
    \`Bar\`.\`name\` AS \`barValue\`,
    \`Baz\`.\`rating\` AS \`bazRating\`
  FROM
    \`Foo\`
    JOIN \`Bar\` ON \`Bar\`.\`color\` = \`Foo\`.\`color\`
    LEFT JOIN \`Baz\` ON \`Baz\`.\`rating\` = \`Bar\`.\`rating\`
  WHERE
    \`Foo\`.\`name\` <> 'Danny'
    AND \`Bar\`.\`rating\` > 50
  `);
  })

  it("will select implicit joined values", async () => {
    class Bar extends Entity {
      name = VarChar();
      foo = One(Foo);
    }
  
    const query = new Query(where => {
      const bar = where(Bar);
  
      return where.get(bar.foo.name)
    });
  
    expect(query).toMatchInlineSnapshot(`
SELECT
  \`Foo\`.\`name\` AS \`1\`
FROM
  \`Bar\`
  LEFT JOIN \`Foo\` ON \`Foo\`.\`id\` = \`Bar\`.\`fooId\`
`);
  })
})
