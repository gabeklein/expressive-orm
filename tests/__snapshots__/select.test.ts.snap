// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joins will select implicit joined values 1`] = `
	SELECT
	  \`Foo\`.\`name\` AS \`1\`
	FROM
	  \`Bar\`
	  LEFT JOIN \`Foo\` ON \`Foo\`.\`id\` = \`Bar\`.\`fooId\`
`;

exports[`joins will select joined values 1`] = `
	SELECT
	  \`Foo\`.\`name\` AS \`fooValue\`,
	  \`Bar\`.\`name\` AS \`barValue\`,
	  \`Baz\`.\`rating\` AS \`bazRating\`
	FROM
	  \`Foo\`
	  JOIN \`Bar\` ON \`Bar\`.\`color\` = \`Foo\`.\`color\`
	  LEFT JOIN \`Baz\` ON \`Baz\`.\`rating\` = \`Bar\`.\`rating\`
	WHERE
	  \`Foo\`.\`name\` <> 'Danny'
	  AND \`Bar\`.\`rating\` > 50
`;

exports[`where.one will limit results 1`] = `
	SELECT
	  \`Foo\`.\`bar\` AS \`bar\`,
	  \`Foo\`.\`baz\` AS \`baz\`
	FROM
	  \`Foo\`
	LIMIT
	  1
`;

exports[`where.get will select a entire entity 1`] = `
	SELECT
	  \`Foo\`.\`bar\` AS \`bar\`,
	  \`Foo\`.\`baz\` AS \`baz\`
	FROM
	  \`Foo\`
`;

exports[`where.get will select a field directly 1`] = `
	SELECT
	  \`Foo\`.\`bar\` AS \`1\`
	FROM
	  \`Foo\`
`;

exports[`where.get will select via map function 1`] = `
	SELECT
	  \`Foo\`.\`bar\` AS \`1\`,
	  \`Foo\`.\`baz\` AS \`2\`
	FROM
	  \`Foo\`
`;

exports[`where.get will select via object 1`] = `
	SELECT
	  \`Foo\`.\`bar\` AS \`bar\`,
	  \`Foo\`.\`baz\` AS \`baz\`
	FROM
	  \`Foo\`
`;
